### DISCUSSION TITLE
Output as HTML

### DISCUSSION THREAD

**Discussion by jeremyzilar** [formfcw/directus-extension-flexible-editor#31] [Q&A] [answered]
I just noticed that TipTap has an option to render the content as HTML as opposed to JSON. There are specific places where we can use that in our set up and was wondering if you'd consider adding that as an option on the field level?
https://tiptap.dev/docs/editor/guide/output

Also, if you can output HTML, is it possible to import HTML into this field? I know it wouldn't build the relations with other items, but being able to drop in a lot of HTML would be amazing. We are migrating to Directus from an old site that has a lot of HTML in the content body.

**Comment by formfcw** [comment, marked as answer]
Storing the editor content as HTML is not planned. Of course, you can render HTML from your JSON content.

For converting HTML to Tiptap/Prosemirror JSON, have a look at https://tiptap.dev/docs/editor/api/utilities/html

**Reply by jeremyzilar** [reply]
Thank you 🙏 I was able to implement that, now trying to figure out how to render the embedded blocks while using `generateHTML`. Do you have any tips? I am wondering if I will need to write an extension to be able to render them

**Reply by m4rrc0** [reply]
@jeremyzilar , have you found a solution?
I am trying to implement a custom extension but I don't seem to find a way to access the data from my relation blocks.

I have something like this:

```
import { generateHTML } from "@tiptap/html";
import { Node } from "@tiptap/core";

const renderToHtml = (contentJson) => {
	// contentJson = {
	// 	type: "doc",
	// 	content: [
	// 		{
	// 			type: "relation-block",
	// 			attrs: {
	// 				id: "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxx",
	// 				junction: "prose_composers_nodes",
	// 				collection: "blocks_heading",
	// 			},
	// 		},
	// 	],
	// };
	const html = generateHTML(contentJson, [
		// other extensions …
		RelationBlock,
	]);

	return html;
};

const RelationBlock = Node.create({
	name: "relation-block",
	renderHTML({ HTMLAttributes }) {
		// HTMLAttributes = {}
		return ["div", HTMLAttributes];
	},
});
```

@formfcw , how are you retrieving your block data in your implementations? I have had a look at your code but I have to admit I don't fully grasp everything you do.

**Reply by jeremyzilar** [reply]
Ultimately, converting our HTML to Tiptap/Prosemirror JSON was too difficult and I had to abandon the idea of using a block editor. The block editor was also too limited for our editorial workflow, and customizing/extending it proved to also be too difficult.

**Reply by m4rrc0** [reply]
Thanks for the prompt response @jeremyzilar and for sharing your experience.

I have worked with other block editors in the past and I must say Tiptap/Prosemirror seems awfully complicated for some basic tasks. From what I read I understand the tool is more focused on the front-end. Backend conversion from JSON to html seems to be an afterthought...

**Reply by formfcw** [reply]
Thanks for sharing your perspectives.

@m4rrc0 Have you read the [README of Flexible Editor Content](https://github.com/formfcw/directus-extension-flexible-editor/blob/main/content/README.md)?

From what I understand you are trying to render html server-side, right? What stack are you using? Basically your code does not look that bad. If you want to resolve a `relation-block` you will need to write your individual render function. What you may have overlooked, is, that you also need to fetch the m2a-data seperately. So in your `relation-block`s `attrs` you will find `id`, `junction` and `collection`. Use these to filter from your existing payload or fetch data from your api.

Hope this helps!

**Reply by formfcw** [reply]
> Ultimately, converting our HTML to Tiptap/Prosemirror JSON was too difficult and I had to abandon the idea of using a block editor. The block editor was also too limited for our editorial workflow, and customizing/extending it proved to also be too difficult.

@jeremyzilar Sorry to hear that. Adding the ability to add custom extensions is on the roadmap, at the same time it is not an easy task as you can imagine.

**Reply by formfcw** [reply]
Please feel free to contribute, as long as you start a discussion first.

**Reply by m4rrc0** [reply]
> @m4rrc0 Have you read the [README of Flexible Editor Content](https://github.com/formfcw/directus-extension-flexible-editor/blob/main/content/README.md)?

Yes. I wouldn't have gone far without it. Thank you @formfcw .

> What you may have overlooked, is, that you also need to fetch the m2a-data seperately.

I do fetch the m2a data and inject it. As you can see in the commented parts of my code, I have some data on the `attrs` key but that data is nowhere to be found in the `renderHTML` method. Is it supposed to be available in the `HTMLAttributes`?? My `HTMLAttributes` is just an empty object...

**Reply by formfcw** [reply]
You need to set the addAttributes method. Have a look at https://tiptap.dev/docs/editor/guide/custom-extensions#attributes

**Reply by formfcw** [reply]
@m4rrc0 have a look at https://github.com/formfcw/directus-extension-flexible-editor/blob/main/src/interface/nodes/relation-block.ts

**Reply by m4rrc0** [reply]
> You need to set the addAttributes method.

Damn... I totally missed that. Thank you @formfcw .

For anyone reading, it seems that `renderHTML` will only receive attributes that have been declared in `addAttributes`.
So the previous code I shared becomes:
```js
import { generateHTML } from "@tiptap/html";
import { Node } from "@tiptap/core";

const renderToHtml = (contentJson) => {
	// contentJson = {
	// 	type: "doc",
	// 	content: [
	// 		{
	// 			type: "relation-block",
	// 			attrs: {
	// 				id: "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxx",
	// 				junction: "prose_composers_nodes",
	// 				collection: "blocks_heading",
	// 			},
	// 		},
	// 	],
	// };
	const html = generateHTML(contentJson, [
		// other extensions …
		RelationBlock,
	]);

	return html;
};

const RelationBlock = Node.create({
	name: "relation-block",
        addAttributes() {
		return {
			id: { default: null },
			junction: { default: null },
			collection: { default: null },
		};
	},
	renderHTML({ HTMLAttributes }) {
		// HTMLAttributes = { id: 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxx', junction: 'prose_composers_nodes', collection: 'heading_blocks' }
                // TODO: create your html elements according to your data
		return ["div", HTMLAttributes];
	},
});
```

**Reply by m4rrc0** [reply]
@formfcw , do you know a decent way of converting normal html to the array notation to be returned by `renderHTML`?
Something like this:
```js
renderHTML({ HTMLAttributes }) {
	return covertHtml(`<div>Hello World</div>`) // converts to ["div", "Hello World"];
},
```

**Reply by formfcw** [reply]
The answer is close 🙈:
> For converting HTML to Tiptap/Prosemirror JSON, have a look at https://tiptap.dev/docs/editor/api/utilities/html

… or regex

**Reply by m4rrc0** [reply]
I may not connect the dots like you do...

> The answer is close 🙈:

I guess you are referring to the `generateJSON` utility? But the JSON returned is nothing like what `renderHTML` expects.

I found this [stackoverflow question](https://stackoverflow.com/questions/74549035/is-it-possible-to-have-a-complete-custom-renderhtml-function-in-tiptap) asking for the same thing, except the answer is... React. And may need a Tiptap Editor instance...? Not sure...

On the other hand, I noticed that returning a string from `renderHTML` actually kind of works... except it escapes the html.

```js
renderHTML({ HTMLAttributes }) {
	return `<div>Hello World</div>` // will return "&lt;div&gt;Hello World&lt;/div&gt;"
}
```

So it seems my options are either:

1. manually parsing html (or JSON with the output of `generateJSON`)
2. using some kind of component library to wrap my html string (and avoid characters being escaped, and process it further if needed)
3. simply decode the output to unescape html characters I need to keep

I need to do some more testing and tinkering but I might just go with the later option and check for code injection on some other level.

If someone lands here looking for more info and I haven't posted further updates, feel free to ask.

**Reply by m4rrc0** [reply]
I think I will be going with a 4th option. 😂
Ultimately, I see little value in processing `relation-block`s with Tiptap since I am only interested in rendering html from them on the server. So I am doing something in those lines to skip Tiptap completely for `relation-block`s:

```js
const flexible_editor_blocks = flexible_editor.content.map((blockRaw) => {
	let block = {};
	if (blockRaw.type === "relation-block") {
		// Do whatever you like however you like without Tiptap getting in the way
		block = normalizeBlock(blockRaw);
	} else {
		// Template is useful to actually render HTML. It seems to render text without this wrapper "doc" type
		const template = { type: "doc", content: [ blockRaw ] };
		block = {
			...blockRaw,
			html: renderToHtml(template),
		};
	}
	return block;
});
```

I don't have enough testing to see if I am missing some important Tiptap feature doing this but at first glance it seems way more satisfying.
